sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/TaskConfigureAwait.cs

```diff
@@ -33,39 +33,30 @@ namespace SonarAnalyzer.Rules.CSharp
     public sealed class TaskConfigureAwait : SonarDiagnosticAnalyzer
     {
         internal const string DiagnosticId = "S3216";
-        private const string MessageFormat =
-            "Add '.ConfigureAwait(false)' to this call to allow execution to continue in any thread.";
+        private const string MessageFormat = "Add '.ConfigureAwait(false)' to this call to allow execution to continue in any thread.";
 
-        private static readonly DiagnosticDescriptor rule =
-            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        private static readonly DiagnosticDescriptor rule = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
 
         public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(rule);
 
-        protected override void Initialize(SonarAnalysisContext context)
-        {
+        protected override void Initialize(SonarAnalysisContext context) =>
             context.RegisterSyntaxNodeActionInNonGenerated(
                 c =>
                 {
-                    if (c.SemanticModel.Compilation.Options.OutputKind != OutputKind.DynamicallyLinkedLibrary)
+                    if (c.Compilation.Options.OutputKind != OutputKind.DynamicallyLinkedLibrary
+                        || !c.Compilation.IsNetFrameworkTarget())
                     {
-                        //this rule only makes sense in libraries
+                        // This rule only makes sense in libraries under .NET Framework
                         return;
                     }
 
-                    var awaitExpression = (AwaitExpressionSyntax)c.Node;
-                    var expression = awaitExpression.Expression;
-                    if (expression == null)
-                    {
-                        return;
-                    }
-
-                    var type = c.SemanticModel.GetTypeInfo(expression).Type;
-                    if (type.DerivesFrom(KnownType.System_Threading_Tasks_Task))
+                    if ((c.Node as AwaitExpressionSyntax).Expression is { } expression
```

Comment:
We should use direct cast here. It is always an `AwaitExpressionSyntax`. If it isn't we would prefer to have a cast exception and not a null reference one.